Typescript: Migrate `@storybook/polymer`

kroeder opened the pull request:
## What I did

Issue https://github.com/storybookjs/storybook/issues/5030

kroeder committed:
wip: start typescript migration of polymer

kroeder committed:
remove exclusion of __tests__

kroeder committed:
add types for polymer/server

kroeder commented on app/polymer/src/server/framework-preset-polymer.ts:
```diff
@@ -10,7 +10,7 @@ export function webpack(config) {
           test: /\.html$/,
           use: [
-            ...config.module.rules[0].use,
+            ...(config.module.rules[0].use as RuleSetUseItem[]),
```
tslint complained about the typing of `use`. I had to cast it but it might be a bug

`rules[0].use` can be ```    type RuleSetUse =
        | RuleSetUseItem
        | RuleSetUseItem[]
        | ((data: any) => RuleSetUseItem | RuleSetUseItem[]);```

or are we 100% certain it is always `RuleSetUseItem[]`? Do we need to add a check here?

ndelangen replied:
We can't be not sure, we should probably check

kroeder committed:
wip: migrate polymer/client to typescript

kroeder commented:
I could use some help migrating the client `storyFn` parts of polymer to typescript 😬 
Can you help or ping someone who can help with that? @ndelangen @shilman

ndelangen committed:
Merge branch 'next' into ts-migration/polymer

# Conflicts:
#	app/polymer/package.json

ndelangen commented:
@daKmoR would you possibly be able to help with this one?

daKmoR commented:
with `storyFn` you mean what the user has in his `foo.stories.js` files right?

then it would be
```ts
import { TemplateResult } from 'lit-html';

export type StoryFnPolymerReturnType = string | Node | TemplateResult;
```

it's actually also "wrong" in `app/web-components` 🙈 
https://github.com/storybookjs/storybook/blob/next/app/web-components/src/client/preview/types.ts#L3

kroeder committed:
wip: proceed ts migration

kroeder committed:
Merge branch 'next' into ts-migration/polymer

kroeder committed:
trigger ci

kroeder commented:
@daKmoR @ndelangen can you review this, please?

ndelangen committed:
Merge branch 'next' into ts-migration/polymer

shilman committed:
Merge branch 'next' into ts-migration/polymer

shilman reviewed (approved):
LGTM. FYI, I'm ok with merging & releasing in 5.3 even tho we're in beta. Are you guys @kroeder @ndelangen @daKmoR ?

ndelangen committed:
Merge branch 'next' into ts-migration/polymer

ndelangen commented:
Let's not take the chance @shilman.

Let's make this a 6.0.0 feature

ndelangen committed:
Merge branch 'next-6.0.0' into ts-migration/polymer

# Conflicts:
#	app/polymer/package.json
#	lib/ui/src/components/sidebar/SidebarHeading.tsx

ndelangen committed:
Merge branch 'next-6.0.0' into ts-migration/polymer

ndelangen committed:
Merge branch 'next-6.0.0' into ts-migration/polymer

gaetanmaisse committed:
fix: clean usage of `@types/webpack-env` dep in all packages

 - `@types/webpack-env` must be defined as `dependencies` in packages that export a type/function using typings from `@types/webpack-env` (it is often `NodeModule`). In that case `/// <reference types="webpack-env" />` is present in some files of the `dist` folder of the lib. Example: `lib/client-api`.
 - `@types/webpack-env` must be defined as `devDependencies` in packages that use its typings internally but does not expose anything related to it. Many addons need the type because they use `module.hot` function. Example: `addons/a11y`
 - `webpack-env` should be removed from "compilerOptions.types" key of `tsconfig.json` of packages that don't need it.

 I also bump `@types/webpack-env` to ^1.15.0 to have the same version all across the monorepo

devtronic committed:
Update 1-Button.stories.ts

The property in the Angular component is lowercase. Without this change, the button example story wont work properly.
(The button text isn't rendered)

ndelangen committed:
Merge branch 'next-6.0.0' into ts-migration/polymer

shilman committed:
Clean usage of `@types/webpack-env` dep in all packages (#9536)

Clean usage of `@types/webpack-env` dep in all packages

ndelangen committed:
Next 6.0.0 (#9212)

Next 6.0.0

ndelangen committed:
Fixed Angular button example story (#9540)

Fixed Angular button example story

shilman committed:
6.0.0-alpha.0 changelog

ndelangen committed:
Merge branch 'next' into pr/kroeder/8102

# Conflicts:
#	app/polymer/package.json
